src/Orleans.ScheduledJobs/LocalScheduledJobManager.cs

```diff
@@ -0,0 +1,473 @@
+
+    /// <summary>
+    /// Attempts to cancel a previously scheduled job.
+    /// </summary>
+    /// <param name="job">The scheduled job to cancel.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation that returns <see langword="true"/> if the job was successfully canceled; otherwise, <see langword="false"/>.</returns>
+    Task<bool> TryCancelScheduledJobAsync(IScheduledJob job);
+}
+
+/// <inheritdoc/>
+internal partial class LocalScheduledJobManager : SystemTarget, ILocalScheduledJobManager, ILifecycleParticipant<ISiloLifecycle>
+{
+    private readonly JobShardManager _shardManager;
+    private readonly IAsyncEnumerable<ClusterMembershipSnapshot> _clusterMembershipUpdates;
+    private readonly ILogger<LocalScheduledJobManager> _logger;
+    private readonly ScheduledJobsOptions _options;
+    private CancellationTokenSource _cts = new();
+    private Task? _listenForClusterChangesTask = null;
+    private Task? _watchForShardtoStartTask = null;
+    private readonly ConcurrentDictionary<DateTimeOffset, ConcurrentDictionary<string, JobShard>> _shardCache = new();
+    private readonly ConcurrentDictionary<string, Task> _runningShards = new();
+    private readonly int MaxJobCountPerShard = 1000;
+    private readonly SemaphoreSlim _jobConcurrencyLimiter;
+
+    private static readonly IDictionary<string, string> EmptyMetadata = new Dictionary<string, string>();
+
+    public LocalScheduledJobManager(
+        JobShardManager shardManager,
+        IClusterMembershipService clusterMembership,
+        IOptions<ScheduledJobsOptions> options,
+        SystemTargetShared shared,
+        ILogger<LocalScheduledJobManager> logger)
+        : base(SystemTargetGrainId.CreateGrainType("scheduledjobs-manager"), shared)
+    {
+        _shardManager = shardManager;
+        _clusterMembershipUpdates = clusterMembership.MembershipUpdates;
+        _logger = logger;
+        _options = options.Value;
+        _jobConcurrencyLimiter = new SemaphoreSlim(_options.MaxConcurrentJobsPerSilo);
+    }
+
+    /// <inheritdoc/>
+    public async Task<IScheduledJob> ScheduleJobAsync(GrainId target, string jobName, DateTimeOffset dueTime, IReadOnlyDictionary<string, string>? metadata = null)
+    {
+        LogSchedulingJob(_logger, jobName, target, dueTime);
+        
+        var key = GetShardKey(dueTime);
+        // Try to get all the available shards for this key
+        var shards = _shardCache.GetOrAdd(key, _ => new ConcurrentDictionary<string, JobShard>());
+        // Find a shard that can accept this job
+        // TODO more efficient lookup
+        foreach (var shard in shards.Select(s => s.Value))
+        {
+            if (!shard.IsComplete && shard.StartTime <= dueTime && shard.EndTime >= dueTime && await shard.GetJobCount() <= MaxJobCountPerShard)
+            {
+                var job = await shard.ScheduleJobAsync(target, jobName, dueTime, metadata);
```

Comment:
We should use a kind of `TryScheduleJobAsync` API so we return `null` when the job wasn't scheduled because the shard is 'complete'.